tests/set_maintenance/test_set_maintenance.py

```diff
@@ -0,0 +1,178 @@
+
+    controller = _run_test(cmd_args, True, repo1, repo2)
+
+    # upload and publish should be called once each
+    assert len(controller.upload_history) == 1
+    assert controller.upload_history[0].repository.id == "redhat-maintenance"
+    assert len(controller.publish_history) == 1
+    assert controller.publish_history[0].repository.id == "redhat-maintenance"
+
+    # logged message should indicate both repos are set to maintenance
+    mock_on_logger.info.assert_called_with(
+        "Setting following repos to Maintenance Mode: \n%s", "repo1\nrepo2"
+    )
+
+
+def test_maintenance_on_with_regex(mock_on_logger):
+    """Test set maintenance by using regex"""
+    repo1 = Repository(id="repo1")
+    repo2 = Repository(id="repo2")
+
+    cmd_args = ["--repo-regex", "repo1"]
+
+    controller = _run_test(cmd_args, True, repo1, repo2)
+
+    # only repo1 should be set to maintenance
+    mock_on_logger.info.assert_called_with(
+        "Setting following repos to Maintenance Mode: \n%s", "repo1"
+    )
+
+
+def test_maintenance_on_with_repo_not_exists():
+    """Set maintenance to non-existed repo in server will fail"""
+    repo1 = Repository(id="repo1")
+    cmd_args = ["--repo-id", "repo2"]
+
+    with pytest.raises(PulpException):
+        _run_test(cmd_args, True, repo1)
+
+
+def test_maintenance_off(mock_off_logger):
+    repo1 = Repository(id="repo1")
+    repo2 = Repository(id="repo2")
+
+    # set maintenance first, get the controller
+    cmd_args = ["--repo-id", "repo1", "repo2"]
+    controller = _run_test(cmd_args, True, repo1, repo2)
+
+    args = [
+        "",
+        "--pulp-url",
+        "http://some.url",
+        "--verbose",
+        "--repo-id",
+        "repo1",
+        "repo2",
+    ]
+    s_m = SetMaintenanceOff()
+
+    # remove repos from maintenance mode by id
+    with patch("sys.argv", args):
+        with _patch_pulp_client(controller.client):
+            s_m.main()
+
+    mock_off_logger.info.assert_called_with(
+        "Following repositories will be removed from Maintenance Mode: \n%s",
+        "repo1\nrepo2",
+    )
+
+    # upload and publish should be called twice eachassert len(controller.upload_history) == 1
+    assert len(controller.upload_history) == 2
+    assert len(controller.publish_history) == 2
+
+
+def test_maintenance_off_with_regex(mock_off_logger):
+    repo1 = Repository(id="repo1")
+    repo2 = Repository(id="repo2")
+
+    # set maintenance first, get the controller
+    cmd_args = ["--repo-id", "repo1", "repo2"]
+    controller = _run_test(cmd_args, True, repo1, repo2)
+
+    args = ["", "--pulp-url", "http://some.url", "--verbose", "--repo-regex", "repo1"]
+    s_m = SetMaintenanceOff()
+
+    # remove repos from maintenance mode by id
+    with patch("sys.argv", args):
+        with _patch_pulp_client(controller.client):
+            s_m.main()
+
+    mock_off_logger.info.assert_called_with(
+        "Following repositories will be removed from Maintenance Mode: \n%s", "repo1"
```

Comment:
I think at least one of these tests (or ideally even all) should get the maintenance report from the fake client and check that it matches the expected outcome, so it's not only testing that it *claimed to* adjust maintenance mode.

Reply:
I thought about it, but since I forget adding maintenance_report in the fake controller, if I want to test it here, I need to access the private attribute _maintenance_report of the fake client, which I shouldn't. 
And think more about it, what the set maintenance commands really do is calling kinds of APIs that pubtools.pulplib provides, it doesn't confirm if they work or not. It's pubtools.pulplib's job to make sure they work.

I can make a pull request to expose the maintenance report in fake controller and test here if you think it would make the tests more reliable.

Reply:
> I thought about it, but since I forget adding maintenance_report in the fake controller, if I want to test it here, I need to access the private attribute _maintenance_report of the fake client, which I shouldn't.

Why do you need an extra method or any private attribute to get the maintenance report from the fake client/controller?  The public get_maintenance_report method ought to work, same as it would on a real client.

> And think more about it, what the set maintenance commands really do is calling kinds of APIs that pubtools.pulplib provides, it doesn't confirm if they work or not. It's pubtools.pulplib's job to make sure they work.

That's true, but it's the job of the code in pubtools-pulp to call those APIs correctly, and it's the job of the tests to verify that it indeed calls those APIs correctly, and the tests currently don't validate that at all.

I can see this by doing a little mutation testing.  I took this PR and made the following small patch:

```
diff --git a/pubtools/_pulp/tasks/set_maintenance/set_maintenance_off.py b/pubtools/_pulp/tasks/set_maintenance/set_maintenance_off.py
index 42a839b..18154a4 100644
--- a/pubtools/_pulp/tasks/set_maintenance/set_maintenance_off.py
+++ b/pubtools/_pulp/tasks/set_maintenance/set_maintenance_off.py
@@ -42,7 +42,7 @@ class SetMaintenanceOff(SetMaintenance):
             for repo_id in to_remove:
                 LOG.info(" - %s", repo_id)
 
-            report = report.remove(to_remove, owner=self.args.owner)
+            #report = report.remove(to_remove, owner=self.args.owner)
 
         return report
```

So, with this patch the maintenance-off command completely fails to remove any repos from maintenance mode.  However, all autotests still pass:

```
$ py.test tests/set_maintenance/test_set_maintenance.py -v
============================================= test session starts =============================================
platform linux -- Python 3.7.4, pytest-5.0.0, py-1.8.0, pluggy-0.12.0 -- /home/rmcgover/dev/python3/bin/python3
cachedir: .pytest_cache
rootdir: /home/rmcgover/src/pubtools-pulp, inifile: tox.ini
plugins: requests-mock-1.6.0, cov-2.7.1
collected 6 items                                                                                             

tests/set_maintenance/test_set_maintenance.py::test_maintenance_on PASSED                               [ 16%]
tests/set_maintenance/test_set_maintenance.py::test_maintenance_on_with_regex PASSED                    [ 33%]
tests/set_maintenance/test_set_maintenance.py::test_maintenance_on_with_repo_not_exists PASSED          [ 50%]
tests/set_maintenance/test_set_maintenance.py::test_maintenance_off PASSED                              [ 66%]
tests/set_maintenance/test_set_maintenance.py::test_maintenance_off_with_regex PASSED                   [ 83%]
tests/set_maintenance/test_set_maintenance.py::test_maintenance_off_with_repo_not_in_maintenance PASSED [100%]

========================================== 6 passed in 0.16 seconds ===========================================
```

This proves that the tests aren't effectively covering the behavior, since they wrongly pass even if the task implementation is completely broken.

The tests also fail to protect against an entire class of other bugs, such as: maintenance-off always unsets maintenance mode for *all* repos instead of the ones selected.

Getting the maintenance report from the client and verifying that the expected changes were made to it, will only cost a little extra code and make the tests much stronger.